sdk/core/core/src/az_span.c

```diff
@@ -131,95 +175,128 @@ AZ_NODISCARD az_result az_span_to_uint32(az_span span, uint32_t* out_number)
 
 AZ_NODISCARD az_result az_span_copy(az_span destination, az_span source, az_span* out_span)
 {
-  AZ_PRECONDITION_VALID_SPAN(destination, 0, true);
-  AZ_PRECONDITION_VALID_SPAN(source, 0, true);
+  AZ_PRECONDITION_NOT_NULL(out_span);
+
   int32_t src_len = az_span_length(source);
+  int32_t dest_capacity = az_span_capacity(destination);
 
-  if (az_span_capacity(destination) < src_len)
+  if (dest_capacity < src_len)
   {
     return AZ_ERROR_INSUFFICIENT_SPAN_CAPACITY;
-  };
+  }
 
   uint8_t* ptr = az_span_ptr(destination);
+  if (ptr == NULL)
+  {
+    *out_span = AZ_SPAN_NULL;
+  }
+  else
+  {
+    //TODO: What should be the behavior if source.ptr is NULL?
+    memmove((void*)ptr, (void const*)az_span_ptr(source), (size_t)src_len);
 
-  memmove((void*)ptr, (void const*)az_span_ptr(source), (size_t)src_len);
-
-  *out_span = az_span_init(ptr, src_len, az_span_capacity(destination));
-
+    *out_span = _az_span_init_unchecked(ptr, src_len, dest_capacity);
+  }
   return AZ_OK;
 }
 
-AZ_NODISCARD AZ_INLINE bool should_encode(uint8_t c)
-{
-  switch (c)
-  {
-    case '-':
-    case '_':
-    case '.':
-    case '~':
-      return false;
-    default:
-      return !(('0' <= c && c <= '9') || ('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'));
-  }
-}
+static const uint8_t _is_ascii_byte_url[256] = {
+  3, 3, 3, 3, 3, 3, 3, 3, // U+00..U+07
+  3, 3, 3, 3, 3, 3, 3, 3, // U+08..U+0F
+  3, 3, 3, 3, 3, 3, 3, 3, // U+10..U+17
+  3, 3, 3, 3, 3, 3, 3, 3, // U+18..U+1F
+  3, 3, 3, 3, 3, 3, 3, 3, // U+20..U+27
+  3, 3, 3, 3, 3, 1, 1, 3, // U+28..U+2F
+  1, 1, 1, 1, 1, 1, 1, 1, // U+30..U+37
+  1, 1, 3, 3, 3, 3, 3, 3, // U+38..U+3F
+  3, 1, 1, 1, 1, 1, 1, 1, // U+40..U+47
+  1, 1, 1, 1, 1, 1, 1, 1, // U+48..U+4F
+  1, 1, 1, 1, 1, 1, 1, 1, // U+50..U+57
+  1, 1, 1, 3, 3, 3, 3, 1, // U+58..U+5F
+  3, 1, 1, 1, 1, 1, 1, 1, // U+60..U+67
+  1, 1, 1, 1, 1, 1, 1, 1, // U+68..U+6F
+  1, 1, 1, 1, 1, 1, 1, 1, // U+70..U+77
+  1, 1, 1, 3, 3, 3, 1, 3, // U+78..U+7F
+  3, 3, 3, 3, 3, 3, 3, 3, // U+80..U+87
+  3, 3, 3, 3, 3, 3, 3, 3, // U+88..U+8F
+  3, 3, 3, 3, 3, 3, 3, 3, // U+90..U+97
+  3, 3, 3, 3, 3, 3, 3, 3, // U+98..U+9F
+  3, 3, 3, 3, 3, 3, 3, 3, // U+A0..U+A7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+A8..U+AF
+  3, 3, 3, 3, 3, 3, 3, 3, // U+B0..U+B7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+B8..U+BF
+  3, 3, 3, 3, 3, 3, 3, 3, // U+C0..U+C7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+C8..U+CF
+  3, 3, 3, 3, 3, 3, 3, 3, // U+D0..U+D7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+D8..U+DF
+  3, 3, 3, 3, 3, 3, 3, 3, // U+E0..U+E7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+E8..U+EF
+  3, 3, 3, 3, 3, 3, 3, 3, // U+F0..U+F7
+  3, 3, 3, 3, 3, 3, 3, 3, // U+F8..U+FF
+};
+
+AZ_NODISCARD AZ_INLINE bool _should_url_encode(uint8_t c) { return _is_ascii_byte_url[c] == 3; }
```

Comment:
Why 1/3 and not 0/1? :)
Given that it is unlikely that URL encoding is bottleneck, is it worth it to unfold a readable and probably more code-size-compact function into an array?

Reply:
Got it why 1/3.

Reply:
I will add a comment to clarify.